[6.x] Fix PHPStan issues

wimski opened the pull request:
- [ ] Added or updated tests
- [ ] Documented user facing changes
- [ ] Updated CHANGELOG.md

spawnia commented:
Before you move any further, merge the latest changes from `master`. There are a ton of issues that are already solved there.

wimski commented:
:astonished: :see_no_evil:

spawnia commented:
I appreciate you putting in the time and effort. I plan to do my part to fix those errors too, but can not promise when I get to it. I will tell you when I start working on it again.

spawnia commented on src/Execution/ModelsLoader/ModelsLoader.php:
```diff
@@ -9,6 +9,10 @@ interface ModelsLoader
     /**
      * Load the result onto the given parent models.
+     *
+     * @template TModel of \Illuminate\Database\Eloquent\Model
+     *
+     * @param EloquentCollection<int, TModel> $parents
```
no need for a generic type there, just inline `TModel`

spawnia commented on src/Execution/ModelsLoader/PaginatedModelsLoader.php:
```diff
@@ -106,6 +106,10 @@ static function (EloquentBuilder $builder, Relation $relation): EloquentBuilder
     /**
      * Use the underlying model to instantiate a relation by name.
+     *
+     * @param EloquentCollection<int, Model>
+     *
+     * @return Relation<Model>
```
Always use FQCN in PHPDocs

spawnia commented on src/Execution/ModelsLoader/PaginatedModelsLoader.php:
```diff
@@ -116,6 +120,10 @@ protected function relationInstance(EloquentCollection $parents): Relation
     /**
      * Return a fresh instance of a query builder for the underlying model.
+     *
+     * @param EloquentCollection<int, Model> $parents
```
```suggestion
     * @param  EloquentCollection<int, Model>  $parents
```

Always use two spaces for `@param`

spawnia commented on src/LighthouseServiceProvider.php:
```diff
@@ -140,6 +140,7 @@ public function boot(ConfigRepository $configRepository): void
 
         $this->loadRoutesFrom(__DIR__ . '/Support/Http/routes.php');
 
+        /** @var ExceptionHandlerContract $exceptionHandler */
         $exceptionHandler = $this->app->make(ExceptionHandlerContract::class);
         // @phpstan-ignore-next-line larastan overly eager assumes this will always be a concrete instance
```
I believe the `@var` tag is unnecessary with the added line below

spawnia commented on src/Pagination/PaginateDirective.php:
```diff
@@ -152,7 +152,9 @@ public function resolveField(FieldValue $fieldValue): callable
                     $query instanceof QueryBuilder || $query instanceof EloquentBuilder || $query instanceof ScoutBuilder || $query instanceof Relation,
                     "The method referenced by the builder argument of the @{$this->name()} directive on {$this->nodeName()} must return a Builder or Relation."
                 );
+                /** @var QueryBuilder|EloquentBuilder<\Illuminate\Database\Eloquent\Model>|ScoutBuilder|Relation<\Illuminate\Database\Eloquent\Model> $query */
             } else {
+                /** @var EloquentBuilder<\Illuminate\Database\Eloquent\Model> $query */
```
Please revert those and let me examine the error

spawnia commented on src/Pagination/PaginatorField.php:
```diff
@@ -8,6 +8,8 @@
 class PaginatorField
 {
     /**
+     * @param LengthAwarePaginator<mixed> $paginator
```
Just to reiterate, all of those class names will need to be FQCN

spawnia commented on src/Pagination/ZeroPageLengthAwarePaginator.php:
```diff
@@ -5,6 +5,9 @@
 use Illuminate\Support\Collection;
 
+/**
+ * @extends LengthAwarePaginator<mixed>
```
FQCN here as well

spawnia commented on src/Schema/Directives/AllDirective.php:
```diff
@@ -60,7 +60,9 @@ public function resolveField(FieldValue $fieldValue): callable
                     $query instanceof QueryBuilder || $query instanceof EloquentBuilder || $query instanceof ScoutBuilder || $query instanceof Relation,
                     "The method referenced by the builder argument of the @{$this->name()} directive on {$this->nodeName()} must return a Builder or Relation."
                 );
+                /** @var QueryBuilder|EloquentBuilder<\Illuminate\Database\Eloquent\Model>|ScoutBuilder|Relation<\Illuminate\Database\Eloquent\Model> $query */
             } else {
+                /** @var EloquentBuilder<\Illuminate\Database\Eloquent\Model> $query */
```
please revert this, I would like to explore different ways of fixing any error here

spawnia commented on src/Schema/Directives/FindDirective.php:
```diff
@@ -35,9 +35,12 @@ public static function definition(): string
     public function resolveField(FieldValue $fieldValue): callable
     {
         return function ($root, array $args, GraphQLContext $context, ResolveInfo $resolveInfo): ?Model {
+            /** @var \Illuminate\Database\Eloquent\Builder<Model> $builder */
```
I feel like this should not be necessary. Please revert and let us explore other options

spawnia commented on src/Schema/Directives/ModifyModelExistenceDirective.php:
```diff
@@ -31,15 +31,16 @@ public function resolveField(FieldValue $fieldValue): callable
         $scopes = $this->directiveArgValue('scopes', []);
 
         return function ($root, array $args, GraphQLContext $context, ResolveInfo $resolveInfo) use ($modelClass, $scopes, $expectsList) {
+            /** @var EloquentBuilder<Model> $builder */
```
should not be necessary

wimski commented:
I can do it right now, but we just have to make sure not to do double work again :slightly_smiling_face:

spawnia commented on src/Schema/Directives/MutationExecutorDirective.php:
```diff
@@ -70,7 +70,10 @@ public function __invoke($parent, $args)
     }
 
     /**
+     * @template TModel of \Illuminate\Database\Eloquent\Model
```
The generic only makes sense when used in multiple places. Think of it like a variable, only makes sense when used multiple times too. Should this be part of the return?

spawnia commented on src/Schema/Directives/RelationDirective.php:
```diff
@@ -179,6 +179,11 @@ protected function paginationDefaultCount(): ?int
     }
 
+    /**
+     * @template TModel of \Illuminate\Database\Eloquent\Model
+     *
+     * @param Relation<TModel> $relation
```
inline generic

spawnia commented on src/Schema/Directives/RelationDirectiveHelpers.php:
```diff
@@ -40,6 +40,7 @@ protected function makeBuilderDecorator($root, array $args, GraphQLContext $cont
             }
             assert($builder instanceof QueryBuilder || $builder instanceof EloquentBuilder);
 
+            /** @var QueryBuilder|EloquentBuilder<\Illuminate\Database\Eloquent\Model> $builder */
```
should not be necessary - perhaps the type of `enhanceBuilder()` is fundamentally flawed?

spawnia commented on src/Schema/Types/LaravelEnumType.php:
```diff
@@ -8,25 +8,28 @@
      * Create a GraphQL enum from a Laravel enum type.
      *
-     * @param  class-string<\BenSampo\Enum\Enum>  $enumClass
+     * @param  class-string<TEnum>|string  $enumClass
```
```suggestion
     * @param  class-string<TEnum>  $enumClass
```

spawnia commented on src/Schema/Types/LaravelEnumType.php:
```diff
@@ -40,6 +43,7 @@ public function __construct(string $enumClass, ?string $name = null)
             throw self::classMustExtendBenSampoEnumEnum($enumClass);
         }
 
+        /** @var class-string<TEnum> $enumClass */
```
should be unnecessary

spawnia commented on src/Schema/Types/LaravelEnumType.php:
```diff
@@ -75,13 +79,22 @@ public static function classMustExtendBenSampoEnumEnum(string $enumClass): \Inva
     }
 
+    /**
+     * @template LocalTValue
+     * @template LocalTEnum of \BenSampo\Enum\Enum<LocalTValue>
+     *
+     * @param LocalTEnum $value
```
no type variables necessary

```suggestion
     * @param \BenSampo\Enum\Enum<mixed> $value
```

spawnia commented on src/Schema/Types/LaravelEnumType.php:
```diff
@@ -116,7 +129,7 @@ protected function deprecationReason(Enum $enum): ?string
     /**
      * TODO remove check and inline when requiring bensampo/laravel-enum:6.
      *
-     * @param  class-string<\BenSampo\Enum\Enum>  $enumClass
+     * @param  class-string  $enumClass
```
feels backwards

spawnia commented on src/Subscriptions/SubscriptionServiceProvider.php:
```diff
@@ -30,6 +30,7 @@ public function register(): void
         $this->app->singleton(BroadcastManager::class);
         $this->app->singleton(SubscriptionRegistry::class);
         $this->app->singleton(StoresSubscriptions::class, static function (Container $app): StoresSubscriptions {
+            /** @var ConfigRepository $configRepository */
```
nope

spawnia commented on src/Support/Contracts/ArgBuilderDirective.php:
```diff
@@ -7,12 +7,10 @@ interface ArgBuilderDirective extends Directive
     /**
      * Add additional constraints to the builder based on the given argument value.
      *
-     * @template TModel of \Illuminate\Database\Eloquent\Model
```
here the generic type makes sense, the method should not change the type of model when returning the eloquent builder

spawnia commented on src/Support/Contracts/FieldBuilderDirective.php:
```diff
@@ -9,10 +9,10 @@ interface FieldBuilderDirective extends Directive
      * Add additional constraints to the builder.
      *
-     * @param  \Illuminate\Database\Query\Builder|\Illuminate\Database\Eloquent\Builder  $builder  the builder used to resolve the field
+     * @param  \Illuminate\Database\Query\Builder|\Illuminate\Database\Eloquent\Builder<\Illuminate\Database\Eloquent\Model>  $builder  the builder used to resolve the field
```
I think Model should be in a type variable here

spawnia commented on src/WhereConditions/WhereConditionsBaseDirective.php:
```diff
@@ -20,7 +20,9 @@ abstract class WhereConditionsBaseDirective extends BaseDirective implements Arg
 
     /**
-     * @param  \Illuminate\Database\Query\Builder|\Illuminate\Database\Eloquent\Builder  $builder  the builder used to resolve the field
+     * @template TModel of \Illuminate\Database\Eloquent\Model
```
inline

spawnia commented on src/WhereConditions/WhereConditionsHandler.php:
```diff
@@ -20,8 +20,11 @@ public function __construct(Operator $operator)
+     *
+     * @param  \Illuminate\Database\Query\Builder|\Illuminate\Database\Eloquent\Builder<TModel>  $builder
      * @param  array<string, mixed>  $whereConditions
+     * @param TModel|null $model
```
to reiterate, needs two spaces

spawnia commented on tests/Integration/Subscriptions/SubscriptionTest.php:
```diff
@@ -91,9 +91,15 @@ public function testSendsSubscriptionChannelInBatchedResponse(): void
         $subscribers = $cache->subscribersByTopic('ON_POST_CREATED');
         $this->assertCount(2, $subscribers);
 
+        /** @var Subscriber $subscriber1 */
```
prefer `assert`

spawnia reviewed (changes requested):
I reviewed most of the changes you made. Most comments apply generally and should be applied in multiple places.

We should also generally avoid `@var`. Prefer `assert()` or add `@phpstan-ignore-next-line` with a descriptive comment if you believe there to be a false-positive.

wimski commented:
I have implemented all your suggestions and fixed more issues. Only 7 issues remain, 6 of which are all to do with `Nuwave\Lighthouse\Support\Contracts\ArgBuilderDirective::handleBuilder` templating.

spawnia commented on src/Execution/ResolveInfo.php:
```diff
@@ -78,14 +78,15 @@ public function wouldEnhanceBuilder(object $builder, array $scopes, $root, array
         foreach ($argumentSet->arguments as $argument) {
             $value = $argument->toPlain();
 
+            /** @var Collection<int, ArgBuilderDirective> $filteredDirectives */
```
Let's avoid `@var`

wimski replied:
```
Parameter #1 $callback of method Illuminate\Support\Collection<int,Nuwave\Lighthouse\Support\Contracts\Directive>::filter()
expects (callable(Nuwave\Lighthouse\Support\Contracts\Directive, int): bool)|null,
Closure(Nuwave\Lighthouse\Support\Contracts\ArgBuilderDirective): bool given.  
```

spawnia replied:
Well then add `@phpstan-ignore-next-line`

spawnia commented on src/Execution/ResolveInfo.php:
```diff
@@ -112,12 +113,13 @@ static function ($value) use (&$builder, $directiveFilter) {
     protected static function wouldApplyArgBuilderDirectives(ArgumentSet $argumentSet, object &$builder, \Closure $directiveFilter = null): bool
     {
         foreach ($argumentSet->arguments as $argument) {
+            /** @var Collection<int, ArgBuilderDirective> $filteredDirectives */
```
avoid

wimski replied:
```
  129    Parameter #1 $callback of method Illuminate\Support\Collection<int,Nuwave\Lighthouse\Support\Contracts\Directive>::filter()
         expects (callable(Nuwave\Lighthouse\Support\Contracts\Directive, int):  bool)|null, 
         Closure(Nuwave\Lighthouse\Support\Contracts\ArgBuilderDirective): bool given.  
```

spawnia commented on src/Schema/Directives/MutationExecutorDirective.php:
```diff
@@ -88,6 +89,10 @@ static function (ArgumentSet $argumentSet) use ($update, $model) {
     /**
      * Prepare the execution function for a mutation on a model.
+     *
+     * @template TModel of \Illuminate\Database\Eloquent\Model
```
inline

spawnia commented on src/Schema/Factories/FieldFactory.php:
```diff
@@ -82,6 +83,7 @@ protected function fieldMiddleware(FieldDefinitionNode $fieldDefinitionNode): ar
             })
             ->all();
 
+        /** @var array<FieldMiddleware> $directiveFieldMiddleware */
```
avoid

wimski replied:
```
Method Nuwave\Lighthouse\Schema\Factories\FieldFactory::fieldMiddleware()
should return array<Nuwave\Lighthouse\Support\Contracts\FieldMiddleware>
but returns array<Nuwave\Lighthouse\Support\Contracts\Directive>.    
```

spawnia commented on tests/Integration/Schema/Directives/CountDirectiveDBTest.php:
```diff
@@ -367,7 +367,7 @@ public function testResolveRelationItemsAndCount(): void
         $user = factory(User::class)->create();
 
-        /** @var Collection<Task> $tasks */
+        /** @var Collection<int, Task> $tasks */
```
fqcn or avoid

wimski replied:
So also use FQCN for inline `@var`? And this is a new rule? Because I see a lot of inline docs without FQCN (example: 2 lines above this issue).

spawnia replied:
The code is not perfect yet. Generally prefer assert, and always use fqcn in phpdocs.

spawnia commented on tests/Integration/Subscriptions/Storage/RedisStorageManagerTest.php:
```diff
@@ -106,7 +106,7 @@ public function testSocketIDStoredOnSubscribe(): void
         $createdSubscriber = $storage->subscribersByTopic('TASK_CREATED')->first();
 
-        $this->assertSame('1234.1234', $createdSubscriber->socket_id);
+        $this->assertSame('1234.1234', $createdSubscriber?->socket_id);
```
```suggestion
        $this->assertNotNull($createdSubscriber);
        $this->assertSame('1234.1234', $createdSubscriber->socket_id);
```

spawnia commented on tests/Integration/Subscriptions/SubscriptionTest.php:
```diff
@@ -148,9 +154,9 @@ public function testWithFieldAlias(): void
                 'lighthouse_subscriptions' => [
                     'version' => 1,
-                    'channel' => $subscriber->channel,
+                    'channel' => $subscriber?->channel,
```
Let's assert `$subscriber` is not null instead

spawnia commented on tests/TestsSerialization.php:
```diff
@@ -45,7 +45,6 @@ public function setUser(?Authenticatable $user): void
     {
-        $config = Container::getInstance()->make(ConfigRepository::class);
-        $config->set('cache.stores.array.serialize', true);
+        Container::getInstance()->make(ConfigRepository::class)->set('cache.stores.array.serialize', true);
```
I prefer it separate

spawnia commented on tests/Unit/Schema/TypeRegistryTest.php:
```diff
@@ -103,7 +103,6 @@ public function testPointToScalarClassThroughDirectiveWithoutNamespace(): void
         ');
 
         $scalarType = $this->typeRegistry->handle($scalarNode);
-        assert($scalarType instanceof ScalarType);
```
Seems necessary:

![image](https://user-images.githubusercontent.com/12158000/220887225-82b53163-ddb8-465f-9293-f1bad2f32bdf.png)

wimski commented on src/Console/DirectiveCommand.php:
```diff
@@ -59,28 +59,28 @@ class DirectiveCommand extends LighthouseGeneratorCommand
      * The implemented interfaces.
      *
-     * @var \Illuminate\Support\Collection<class-string>
+     * @var \Illuminate\Support\Collection<int, string>
```
Shouldn't `class-string`s be FQCN? The _short names_ which are pushed in the `implements` property are just strings, because the can't be resolved as actual classes anymore.

spawnia replied:
You are right, nevermind. Thanks for fixing this.

wimski commented on src/Execution/ModelsLoader/ModelsLoader.php:
```diff
@@ -9,6 +9,10 @@ interface ModelsLoader
     /**
      * Load the result onto the given parent models.
+     *
+     * @template TModel of \Illuminate\Database\Eloquent\Model
+     *
+     * @param  \Illuminate\Database\Eloquent\Collection<int, TModel>  $parents
```
```
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   tests/Integration/Execution/DataLoader/PaginatedRelationLoaderTest.php                                                                                                                            
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  37     Parameter #1 $parents of method Nuwave\Lighthouse\Execution\ModelsLoader\PaginatedModelsLoader::load() expects Illuminate\Database\Eloquent\Collection<int, Illuminate\Database\Eloquent\Model>,  
         Illuminate\Database\Eloquent\Collection<int, Tests\Utils\Models\User> given.                                                                                                                      
         💡 Template type TModel on class Illuminate\Database\Eloquent\Collection is not covariant. Learn more: https://phpstan.org/blog/whats-up-with-template-covariant                                  
  79     Parameter #1 $parents of method Nuwave\Lighthouse\Execution\ModelsLoader\PaginatedModelsLoader::load() expects Illuminate\Database\Eloquent\Collection<int, Illuminate\Database\Eloquent\Model>,  
         Illuminate\Database\Eloquent\Collection<int, Tests\Utils\Models\User> given.                                                                                                                      
         💡 Template type TModel on class Illuminate\Database\Eloquent\Collection is not covariant. Learn more: https://phpstan.org/blog/whats-up-with-template-covariant                                  
  109    Parameter #1 $parents of method Nuwave\Lighthouse\Execution\ModelsLoader\PaginatedModelsLoader::load() expects Illuminate\Database\Eloquent\Collection<int, Illuminate\Database\Eloquent\Model>,  
         Illuminate\Database\Eloquent\Collection<int, Tests\Utils\Models\User> given.                                                                                                                      
         💡 Template type TModel on class Illuminate\Database\Eloquent\Collection is not covariant. Learn more: https://phpstan.org/blog/whats-up-with-template-covariant                                  
  116    Parameter #1 $parents of method Nuwave\Lighthouse\Execution\ModelsLoader\PaginatedModelsLoader::load() expects Illuminate\Database\Eloquent\Collection<int, Illuminate\Database\Eloquent\Model>,  
         Illuminate\Database\Eloquent\Collection<int, Tests\Utils\Models\User> given.                                                                                                                      
         💡 Template type TModel on class Illuminate\Database\Eloquent\Collection is not covariant. Learn more: https://phpstan.org/blog/whats-up-with-template-covariant                                  
  158    Parameter #1 $parents of method Nuwave\Lighthouse\Execution\ModelsLoader\PaginatedModelsLoader::load() expects Illuminate\Database\Eloquent\Collection<int, Illuminate\Database\Eloquent\Model>,  
         Illuminate\Database\Eloquent\Collection<int, Tests\Utils\Models\User> given.                                                                                                                      
         💡 Template type TModel on class Illuminate\Database\Eloquent\Collection is not covariant. Learn more: https://phpstan.org/blog/whats-up-with-template-covariant                                  
  189    Parameter #1 $parents of method Nuwave\Lighthouse\Execution\ModelsLoader\PaginatedModelsLoader::load() expects Illuminate\Database\Eloquent\Collection<int, Illuminate\Database\Eloquent\Model>,  
         Illuminate\Database\Eloquent\Collection<int, Tests\Utils\Models\Task> given.                                                                                                                      
         💡 Template type TModel on class Illuminate\Database\Eloquent\Collection is not covariant. Learn more: https://phpstan.org/blog/whats-up-with-template-covariant                                  
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
```

spawnia replied:
Interesting, I guess this has to do with covariance/contravariance.

wimski commented on src/Schema/Factories/FieldFactory.php:
```diff
@@ -73,6 +73,7 @@ public function handle(FieldValue $fieldValue): array
      */
     protected function fieldMiddleware(FieldDefinitionNode $fieldDefinitionNode): array
     {
+        /** @var array<FieldMiddleware> $globalFieldMiddleware */
```
```
Method Nuwave\Lighthouse\Schema\Factories\FieldFactory::fieldMiddleware()
should return array<Nuwave\Lighthouse\Support\Contracts\FieldMiddleware>
but returns array<Nuwave\Lighthouse\Support\Contracts\Directive>.    
```

wimski replied:
Should I add line for this error in the `phpstan.neon` file, limited to the FieldFactory path?

spawnia replied:
No, add `@phpstan-ignore-next-line` with a description where the error occurs.

spawnia commented on tests/Unit/Schema/TypeRegistryTest.php:
```diff
@@ -141,8 +142,7 @@ interface Nameable {
         ');
 
         $interfaceType = $this->typeRegistry->handle($interfaceNode);
-        assert($interfaceType instanceof InterfaceType);
```
Why change this?

wimski replied:
:shrug: reverted

spawnia commented on src/Execution/ResolveInfo.php:
```diff
@@ -91,9 +93,11 @@ protected static function applyArgBuilderDirectives(ArgumentSet $argumentSet, ob
                 ->filter(Utils::instanceofMatcher(ArgBuilderDirective::class));
 
             if (null !== $directiveFilter) {
+                // @phpstan-ignore-next-line
```
```suggestion
                // @phpstan-ignore-next-line PHPStan does not get this list is filtered for ArgBuilderDirective
```

spawnia commented on src/Execution/ResolveInfo.php:
```diff
@@ -91,9 +93,11 @@ protected static function applyArgBuilderDirectives(ArgumentSet $argumentSet, ob
                 $filteredDirectives = $filteredDirectives->filter($directiveFilter);
             }
 
+            // @phpstan-ignore-next-line
```
```suggestion
            // @phpstan-ignore-next-line PHPStan does not get this list is filtered for ArgBuilderDirective
```

spawnia commented on src/Execution/ResolveInfo.php:
```diff
@@ -123,6 +127,7 @@ protected static function wouldApplyArgBuilderDirectives(ArgumentSet $argumentSe
                 ->filter(Utils::instanceofMatcher(ArgBuilderDirective::class));
 
             if (null !== $directiveFilter) {
+                // @phpstan-ignore-next-line
```
```suggestion
                // @phpstan-ignore-next-line PHPStan does not get this list is filtered for ArgBuilderDirective
```

spawnia commented:
Thanks! I just checked out your branch and am looking at the remaining issues.
